Sources/Basics/FileSystem/AsyncFileSystem.swift

```diff
@@ -0,0 +1,207 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+#if swift(>=5.5.2)
+import _Concurrency
+
+import class TSCBasic.BufferedOutputByteStream
+import class TSCBasic.FileLock
+import enum TSCBasic.FileMode
+import protocol TSCBasic.FileSystem
+import protocol TSCBasic.WritableByteStream
+import struct TSCBasic.AbsolutePath
+import struct TSCBasic.ByteString
+import struct TSCBasic.FileInfo
+import struct TSCBasic.FileSystemError
+
+/// Abstracted access to file system operations.
+///
+/// This actor is used to allow most of the codebase to interact with an
+/// asynchronous filesystem interface, while still allowing clients to transparently
+/// substitute a virtual file system or redirect file system operations.
+public actor AsyncFileSystem: Actor {
```

Comment:
should all the methods here be marked as async? a true FS is async in nature and I think we can prepare the API so that all the call sites do the right thing from the get go, event tho the existing underlying FS is blocking

Reply:
All of the methods (except `nonisolated`, which we don't have here) on actors get implicit `async` requirement on them when called outside of the actor.

Or do you mean that they should be explicitly `async` even within the actor? That would add `await` suspension points within implementations even when it calls its own methods. This has performance impact, but no other benefits, as far as I understand, because neither Foundation nor SwiftNIO support true async I/O on any platforms.

Reply:
right, my point is that at some point we will have a true FS impl (in Foundation, or SwiftNIO, or something new). your point about implicit async for actors is a good one, I forgot about that

Reply:
I think if/when underlying FS becomes truly async, it wouldn't be source-breaking to add explicit `async` effect modifiers on these methods, since the client code using them would already have `await` in front of calls to them anyway. Only code within `AsyncFileSystem` would change, which makes that API impact quite isolated.

Reply:
sounds good. lets just a add a few tests and get this merged?